### ISSUE TITLE
Allow Test Classes to observe test errors

### ISSUE THREAD

**Issue by RichiCoder1** [RichiCoder1/Browser.xUnit#2]
It is desirable that test classes be able to be informed of test errors _before_ test class disposal to allow things like dumps and screenshotting. Current xUnit exposes `BeforeAfterTestAttribute`, but this attribute provides neither guarantees on when it will run nor does it allow `After` to know if the test succeeded or failed.
# Proposed Solutions:

@pawelpabich has proposed the addition of two (or technically, three) new interfaces that a test class can implement:
### INeedToKnowTestSuccess

``` csharp
    public interface INeedToKnowTestSuccess : INeedToKnowTestResult
    {
        void Handle(TestPassed result);
    }
```
### INeedToKnowTestFailure

``` csharp
    public interface INeedToKnowTestFailure : INeedToKnowTestResult
    {
        void Handle(TestFailed result);
    }
```
### INeedToKnowTestResult

``` csharp
    public interface INeedToKnowTestResult
    {
    }
```

These interfaces, if they exist, would be invoked after the Test Method has been invoked but before any other actions.

**Comment by RichiCoder1** [comment]
After actually implementing this, I'm questioning the value of INeedToKnowTestSuccess and TestResult. I'd be ok with adding handlers higher up so you can do something similar to [BeforeAfterAssemblyTests], however. Is there any reason the Test Class itself should be interested in successes @pawelpabich?

**Comment by RichiCoder1** [comment]
In other news, I also tweaked the INeedToKnowTestFailure interface. It now looks like below:

``` csharp
public interface INeedToKnowTestFailure
{
    Task HandleFailureAsync(ITest test, Exception exception);
}
```

Using TestFailed, for various reasons, doesn't make sense.

**Comment by pawelpabich** [comment]
@RichiCoder1 yeah, INeedToKnowTestFailure is the only critical one. Is ITest an instance of the test before it gets disposed ?
